File: scala/71.scala
import scala.collection.mutable.Stack
object Solution {
    def simplifyPath(path: String): String = {
        val tokens = path.split('/')
        val stack = Stack[String]()
        for (token <- tokens) {
            token match {
                case "/" | "." | "" => ()
                case ".." => {
                    if (!stack.isEmpty) stack.pop()
                }
                case _ => stack.push(token)
            }
        }       
        "/" + stack.reverse.mkString("/")
    }
}

File: leetcodeJava/71.java
package leetcodeJava;

class Solution {
    public String simplifyPath(String path) {
        var dirs = path.split("/");
        var stack = new Stack<String>();
        for (var dir : dirs) {
            if (dir.equals("..")) {
                if (!stack.empty()) stack.pop();
            } else if (!dir.equals(".") && !dir.equals("")) {
                stack.push(dir);
            }
        }
        var sb = new StringBuilder();
        while(!stack.empty()) {
            sb.insert(0, "/" + stack.pop());
        }
        var res = sb.toString();
        if (res.length() == 0) return "/";
        return res;
    }
}
